docs/developer-docs/build/backend/reproducible-builds.md

```diff
@@ -64,46 +64,55 @@ You should communicate all of these to your user in the instructions. Ideally, d
 
     FROM ubuntu:22.04
 
+    ENV NVM_DIR=/root/.nvm
+    ENV NVM_VERSION=v0.39.1
+    ENV NODE_VERSION=18.1.0
+
+    ENV RUSTUP_HOME=/opt/rustup
+    ENV CARGO_HOME=/opt/cargo
+    ENV RUST_VERSION=1.62.0
+    ENV IC_CDK_OPTIMIZER_VERSION=0.3.4
+
+    ENV DFX_VERSION=0.11.0
+
     # Install a basic environment needed for our build tools
-    RUN \
-        apt -yq update && \
+    RUN apt -yq update && \
         apt -yqq install --no-install-recommends curl ca-certificates \
-            build-essential pkg-config libssl-dev llvm-dev liblmdb-dev clang cmake
+            build-essential pkg-config libssl-dev llvm-dev liblmdb-dev clang cmake rsync
 
     # Install Node.js using nvm
-    # Specify the Node version
-    ENV NODE_VERSION=18.1.0
-    RUN curl --fail -sSf https://raw.githubusercontent.com/creationix/nvm/v0.39.1/install.sh | bash
-    ENV NVM_DIR=/root/.nvm
-    RUN . "$NVM_DIR/nvm.sh" && nvm install ${NODE_VERSION}
-    RUN . "$NVM_DIR/nvm.sh" && nvm use v${NODE_VERSION}
-    RUN . "$NVM_DIR/nvm.sh" && nvm alias default v${NODE_VERSION}
-    ENV PATH="/root/.nvm/versions/node/v${NODE_VERSION}/bin/:${PATH}"
-
-    # Install Rust and Cargo in /opt
-    # Specify the Rust toolchain version
-    ARG rust_version=1.60.0
-    ENV RUSTUP_HOME=/opt/rustup \
-        CARGO_HOME=/opt/cargo \
-        PATH=/opt/cargo/bin:$PATH
+    ENV PATH="/root/.nvm/versions/node/v${NODE_VERSION}/bin:${PATH}"
+    RUN curl --fail -sSf https://raw.githubusercontent.com/creationix/nvm/${NVM_VERSION}/install.sh | bash
+    RUN . "${NVM_DIR}/nvm.sh" && nvm install ${NODE_VERSION}
+    RUN . "${NVM_DIR}/nvm.sh" && nvm use v${NODE_VERSION}
+    RUN . "${NVM_DIR}/nvm.sh" && nvm alias default v${NODE_VERSION}
+
+    # Install Rust and Cargo
+    ENV PATH=/opt/cargo/bin:${PATH}
     RUN curl --fail https://sh.rustup.rs -sSf \
-            | sh -s -- -y --default-toolchain ${rust_version}-x86_64-unknown-linux-gnu --no-modify-path && \
-        rustup default ${rust_version}-x86_64-unknown-linux-gnu && \
+            | sh -s -- -y --default-toolchain ${RUST_VERSION}-x86_64-unknown-linux-gnu --no-modify-path && \
+        rustup default ${RUST_VERSION}-x86_64-unknown-linux-gnu && \
         rustup target add wasm32-unknown-unknown
-    RUN cargo install ic-cdk-optimizer
+    RUN cargo install --version ${IC_CDK_OPTIMIZER_VERSION} ic-cdk-optimizer
 
-    # Install dfx; the version is picked up from the DFX_VERSION environment variable
-    ENV DFX_VERSION=0.11.0
+    # Install dfx
     RUN sh -ci "$(curl -fsSL https://smartcontracts.org/install.sh)"
 
     COPY . /canister
     WORKDIR /canister
 
-    RUN npm ci # if `package.json` and `package-lock.json` is available in your canister directory
+An example script to build a Rust project looks as follows:
```

Comment:
Minor: maybe put this in the Dockerfile and comment it out?

Reply:
I believe that a separate build script is useful because it can also be specified as a custom build script in `dfx.json`. I've added some comments on that.